### PR TITLE
Probe + Launcher (with associated events/transformsync changes)

### PR DESCRIPTION
**Pull request by misternebula**
this one's a big one 🥴

### CODE REVIEW

File: QSB/Tools/PlayerToolsManager.cs [hunk 1]
```diff
@@ -27,6 +28,15 @@ public static void Init(Transform camera)
             CreateFlashlight();
             CreateSignalscope();
             CreateTranslator();
+            CreateProbeLauncher();
+        }
+
+        public static void CreateProbe(Transform body, uint id)
```
Comment (misternebula) [inline QSB/Tools/PlayerToolsManager.cs:34]: This method is here for future functionality.

File: QSB/Events/EventHandler.cs [hunk 2]
```diff
@@ -44,19 +44,15 @@ private void OnServerReceiveMessage(EventMessage message)
 
         private void OnClientReceiveMessage(EventMessage message)
         {
-            if (message.SenderId == PlayerRegistry.LocalPlayer.NetId)
-            {
-                return;
-            }
```
Comment (amazingalek) [inline QSB/Events/EventHandler.cs:50]: should still keep this?

File: QSB/Events/EventHandler.cs [hunk 3]
```diff
@@ -67,20 +63,65 @@ private void OnClientReceiveMessage(EventMessage message)
...
+                    break;
+                case EventType.LaunchProbe:
+                    player.UpdateState(State.ProbeActive, true);
+                    break;
+            }
+            if (message.SenderId == PlayerRegistry.LocalPlayer.NetId)
+            {
+                return;
+            }
+            switch ((EventType)message.EventType)
```
Comment (amazingalek) [inline QSB/Events/EventHandler.cs:93]: one switch was ugly enough, why are there now two? 😂

File: QSB/TransformSync/TransformSync.cs [hunk 4]
```diff
@@ -80,18 +82,17 @@ private void Update()
...
-                DebugLog.ToConsole($"Error - Player ID {PlayerRegistry.LocalPlayer.NetId}'s reference sector is null!", MessageType.Error);
-            }
+            var sectorTransform = PlayerRegistry.GetPlayer(GetAttachedNetId()).ReferenceSector;
 
             if (hasAuthority) // If this script is attached to the client's own body on the client's side.
             {
                 transform.position = sectorTransform.InverseTransformPoint(SyncedTransform.position);
                 transform.rotation = sectorTransform.InverseTransformRotation(SyncedTransform.rotation);
+
+                if (Override())
```
Comment (amazingalek) [inline QSB/TransformSync/TransformSync.cs:92]: I suggest you make Override into a virtual and empty Method that only the probe overrides, and that's called from here. Also rename it to something else.

Comment (amazingalek) [reply QSB/TransformSync/TransformSync.cs:92]: That way you don't have to make the other (not probe) TransformSyncs uglier

File: QSB/TransformSync/TransformSync.cs [hunk 5]
```diff
@@ -108,6 +111,11 @@ private void Update()
 
                     SyncedTransform.localPosition = Vector3.SmoothDamp(SyncedTransform.localPosition, transform.position, ref _positionSmoothVelocity, SmoothTime);
                     SyncedTransform.localRotation = QuaternionHelper.SmoothDamp(SyncedTransform.localRotation, transform.rotation, ref _rotationSmoothVelocity, Time.deltaTime);
+
+                    if (Override())
```
Comment (amazingalek) [inline QSB/TransformSync/TransformSync.cs:115]: Ah, there are two of these. Hmmm. I suggest we find a way to properly teleport the probe. Scratch the above comment.

File: QSB/TransformSync/TransformSync.cs [hunk 6]
```diff
@@ -80,18 +82,17 @@ private void Update()
...
             }
 
-            // Get which sector should be used as a reference point
-            var sectorTransform = PlayerRegistry.LocalPlayer.ReferenceSector;
-
-            if (sectorTransform == null)
-            {
-                DebugLog.ToConsole($"Error - Player ID {PlayerRegistry.LocalPlayer.NetId}'s reference sector is null!", MessageType.Error);
-            }
+            var sectorTransform = PlayerRegistry.GetPlayer(GetAttachedNetId()).ReferenceSector;
```
Comment (amazingalek) [inline QSB/TransformSync/TransformSync.cs:85]: This fixes transformsync of the player but not ships etc. #115 does that.

File: QSB/Events/EventListener.cs [hunk 7]
```diff
@@ -1,40 +1,84 @@
...
+                          let normalParams = m.GetParameters()
+
+                          where typeParams.Length == typeCount
+
+                          select m;
+
+            return methods.Single();
+        }
+
+        // this is the file i hate most in QSB. it should burn forever.
```
Comment (amazingalek) [inline QSB/Events/EventListener.cs:82]: It needs improvements... :P

File: QSB/Events/EventHandler.cs (hunk 2 above)
Comment (misternebula) [reply QSB/Events/EventHandler.cs:50]: I did, it just moved to below the first switch

**Pull request merged**